src/IO/ReadBufferFromBlobStorage.cpp

```diff
@@ -0,0 +1,164 @@
+    // TODO: shall this notation be used in all constructors?
+    : SeekableReadBuffer(nullptr, 0)
+    , blob_container_client(blob_container_client_)
+    , path(path_)
+    , max_single_read_retries(max_single_read_retries_)
+    , tmp_buffer_size(tmp_buffer_size_) // NOTE: field used only here in the constructor
+    , use_external_buffer(use_external_buffer_)
+    , read_until_position(read_until_position_)
+{
+    if (!use_external_buffer)
+    {
+        tmp_buffer.resize(tmp_buffer_size);
+        data_ptr = tmp_buffer.data();
+        data_capacity = tmp_buffer_size;
+    }
+}
+
+
+bool ReadBufferFromBlobStorage::nextImpl()
+{
+    // TODO: is this "stream" approach better than a single DownloadTo approach (last commit 90fc230c4dfacc1a9d50d2d65b91363150caa784) ?
+
+    if (read_until_position)
+    {
+        if (read_until_position == offset)
+            return false;
+
+        if (read_until_position < offset)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Attempt to read beyond right offset ({} > {})", offset, read_until_position - 1);
+    }
+
+    if (!initialized)
+        initialize();
+
+    if (use_external_buffer)
+    {
+        data_ptr = internal_buffer.begin();
+        data_capacity = internal_buffer.size();
+    }
+
+    size_t to_read_bytes = std::min(total_size - offset, data_capacity);
+    size_t bytes_read = 0;
+
+    auto sleep_time_with_backoff_milliseconds = std::chrono::milliseconds(100);
+    for (size_t i = 0; i < max_single_read_retries; ++i) // TODO: Is this try part necessary with Azure reliable streams?
+    {
+        try
+        {
+            bytes_read = data_stream->ReadToCount(reinterpret_cast<uint8_t *>(data_ptr), to_read_bytes);
+            break;
+        }
+        catch (const Azure::Storage::StorageException & e)
+        {
+            LOG_INFO(log, "Exception caught during Azure Read for file {} : {}", path, e.Message);
+
+            std::this_thread::sleep_for(sleep_time_with_backoff_milliseconds);
+            sleep_time_with_backoff_milliseconds *= 2;
+            initialized = false;
+            initialize();
+        }
+    }
+
+    if (bytes_read == 0)
+        return false;
+
+    BufferBase::set(data_ptr, bytes_read, 0);
+    offset += bytes_read;
+
+    return true;
+}
+
+
+off_t ReadBufferFromBlobStorage::seek(off_t offset_, int whence)
+{
+    if (initialized)
+        throw Exception("Seek is allowed only before first read attempt from the buffer.", ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
+
+    if (whence != SEEK_SET)
+        throw Exception("Only SEEK_SET mode is allowed.", ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
+
+    if (offset_ < 0)
+        throw Exception("Seek position is out of bounds. Offset: " + std::to_string(offset_), ErrorCodes::SEEK_POSITION_OUT_OF_BOUND);
+
+    offset = offset_;
+
+    return offset;
+}
+
+
+off_t ReadBufferFromBlobStorage::getPosition()
+{
+    // TODO: why is it `return offset - available()` in S3?
```

Comment:
Because in `nextImpl()` `offset` is incremented with `bytes_read`, but we actually did not yet read it, so need to subtract `available()`.

Reply:
Hm, considering your explanation, wouldn't it rather be something like `return offset()` or `return impl->offset()` then (underneath `pos - working_buffer.begin()`, like in ReadBufferFromMemory.h) ?
Using `offset - available()`, one gets `offset - (working_buffer.end() - pos) = offset + pos - working_buffer.end()`, which doesn't make sense to me.
The explanation for `getPosition()` is "Offset from the begin of the underlying buffer / file corresponds to the buffer current position", but with the initial `offset = 0` and `pos = working_buffer.begin()` we get `getPosition() = (-1) * working_buffer.size()`.

Reply:
>  but with the initial offset = 0 and pos = working_buffer.begin() we get getPosition() = (-1) * working_buffer.size()

No, this will not happen, because with `offset = 0`, i.e. before we read anything we have `working_buffer.begin() == working_buffer.end()` so `getPosition()` will return `0`.

Reply:
In this case I should use `offset - available()` here too, right? I also increment `offset` with `bytes_read` after data fetch, but before a read.